fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/StorageAdapter.java

```diff
@@ -0,0 +1,389 @@
+     * @return the subspace
+     */
+    @Nonnull
+    Subspace getSubspace();
+
+    /**
+     * Gets the subspace that contains the data for this object.
+     * <p>
+     * This subspace represents the portion of the keyspace dedicated to storing the actual data, as opposed to metadata
+     * or other system-level information.
+     * @return the subspace containing the data, which is guaranteed to be non-null
+     */
+    @Nonnull
+    Subspace getDataSubspace();
+
+    /**
+     * Get the on-write listener.
+     * @return the on-write listener.
+     */
+    @Nonnull
+    OnWriteListener getOnWriteListener();
+
+    /**
+     * Get the on-read listener.
+     * @return the on-read listener.
+     */
+    @Nonnull
+    OnReadListener getOnReadListener();
+
+    /**
+     * Asynchronously fetches a node from a specific layer, identified by its primary key.
+     * <p>
+     * The fetch operation is performed within the scope of the provided {@link ReadTransaction}, ensuring a consistent
+     * view of the data. The returned {@link CompletableFuture} will be completed with the node once it has been
+     * retrieved from the underlying data store.
+     * @param readTransaction the {@link ReadTransaction} context for this read operation
+     * @param storageTransform an affine vector transformation operator that is used to transform the fetched vector
+     *        into the storage space that is currently being used
+     * @param layer the layer from which to fetch the node
+     * @param primaryKey the {@link Tuple} representing the primary key of the node to retrieve
+     * @return a non-null {@link CompletableFuture} which will complete with the fetched {@code Node<N>}.
+     */
+    @Nonnull
+    CompletableFuture<Node<N>> fetchNode(@Nonnull ReadTransaction readTransaction,
+                                         @Nonnull AffineOperator storageTransform,
+                                         int layer,
+                                         @Nonnull Tuple primaryKey);
+
+    /**
+     * Writes a node and its neighbor changes to the data store within a given transaction.
+     * <p>
+     * This method is responsible for persisting the state of a {@link Node} and applying any modifications to its
+     * neighboring nodes as defined in the {@code NeighborsChangeSet}. The entire operation is performed atomically as
+     * part of the provided {@link Transaction}.
+     * @param transaction the non-null transaction context for this write operation.
+     * @param quantizer the quantizer to use
+     * @param node the non-null node to be written to the data store.
+     * @param layer the layer index where the node resides.
+     * @param changeSet the non-null set of changes describing additions or removals of
+     *        neighbors for the given {@link Node}.
+     */
+    void writeNode(@Nonnull Transaction transaction, @Nonnull Quantizer quantizer, @Nonnull Node<N> node, int layer,
```

Comment:
FDB `set`s generally are, but is this guaranteed to be non-blocking, and that's why it doesn't return a future? That would make sense, but just to clarify that that's the idea